File: arai60/convert-sorted-array-to-binary-search-tree/step3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def build_bst(left: int, right: int) -> Optional[TreeNode]:
            if right - left == 0: return None
            mid = (left + right) // 2
            root = TreeNode(nums[mid])
            root.left = build_bst(left, mid)
            root.right = build_bst(mid + 1, right)
            return root

        return build_bst(0, len(nums))

File: arai60/validate-binary-search-tree/step3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isValidBSTHelper(root, numeric_limits<long long>::min(), numeric_limits<long long>::max());
    }

private:
    bool isValidBSTHelper(TreeNode* root, long long lower_bound, long long upper_bound) {
        if (!root) return true;
        if (!(lower_bound < root->val && root->val < upper_bound)) return false;
        return isValidBSTHelper(root->left, lower_bound, root->val) && isValidBSTHelper(root->right, root->val, upper_bound);
    }
};